compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1017,7 +1018,41 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             return Ok(cycle_result);
         }
 
-        let (result, dep_node) = self.in_task(|this| this.evaluate_stack(&stack));
+        let (result, dep_node) = self.in_task(|this| {
+            let mut result = this.evaluate_stack(&stack)?;
+
+            // if evaluate failed, try to normalize the predicate and evaluate again.
+            if EvaluationResult::EvaluatedToErr == result
+                && fresh_trait_pred.has_projections()
+                && fresh_trait_pred.is_global()
+            {
+                let mut nested_obligations = Vec::new();
+                let predicate = try_normalize_with_depth_to(
+                    this,
+                    param_env,
+                    obligation.cause.clone(),
+                    obligation.recursion_depth + 1,
+                    fresh_trait_pred,
+                    &mut nested_obligations,
+                );
+                if predicate != fresh_trait_pred {
+                    let obligation = obligation.with(this.tcx(), predicate);
+                    let nested_obligations_result = nested_obligations
+                        .into_iter()
+                        .map(|obligation| {
+                            this.evaluate_predicate_recursively(stack.list(), obligation)
+                        })
+                        .collect::<Result<Vec<_>, _>>()?;
+                    if nested_obligations_result.iter().all(|result| *result == EvaluatedToOk) {
```

Comment:
don't check for `EvaluatedToOk`, instead merge all evaluation results.

Checking for `EvaluatedToOk` for behavior is kinda bad as it causes code to break the moment it adds a lifetime parameter, evne if it is fully unconstrained.